File: 351004/Dashkevich/REST_lab/src/main/java/com/adashkevich/rest/lab/service/MarkerService.java
package com.adashkevich.rest.lab.service;

import com.adashkevich.rest.lab.dto.request.MarkerRequestTo;
import com.adashkevich.rest.lab.dto.response.MarkerResponseTo;
import com.adashkevich.rest.lab.exception.ConflictException;
import com.adashkevich.rest.lab.exception.NotFoundException;
import com.adashkevich.rest.lab.mapper.MarkerMapper;
import com.adashkevich.rest.lab.model.Marker;
import com.adashkevich.rest.lab.model.News;
import com.adashkevich.rest.lab.repository.MarkerRepository;
import com.adashkevich.rest.lab.repository.NewsRepository;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Service
public class MarkerService {
    private final MarkerRepository repo;
    private final NewsRepository newsRepo;
    private final MarkerMapper mapper;

    public MarkerService(MarkerRepository repo, NewsRepository newsRepo, MarkerMapper mapper) {
        this.repo = repo;
        this.newsRepo = newsRepo;
        this.mapper = mapper;
    }

    public MarkerResponseTo create(MarkerRequestTo dto) {
        ensureNameUnique(dto.name, null);

        Marker entity = mapper.toEntity(dto);
        Marker saved = repo.save(entity);
        return mapper.toResponse(saved);
    }

    public List<MarkerResponseTo> getAll() {
        return repo.findAll().stream().map(mapper::toResponse).toList();
    }

    public MarkerResponseTo getById(Long id) {
        Marker m = repo.findById(id)
                .orElseThrow(() -> new NotFoundException("Marker not found", "40430"));
        return mapper.toResponse(m);
    }

    public MarkerResponseTo update(Long id, MarkerRequestTo dto) {
        Marker existing = repo.findById(id)
                .orElseThrow(() -> new NotFoundException("Marker not found", "40430"));

        ensureNameUnique(dto.name, id);

        existing.setName(dto.name);
        Marker updated = repo.update(id, existing);
        return mapper.toResponse(updated);
    }

    public void delete(Long id) {
        Marker marker = repo.findById(id)
                .orElseThrow(() -> new NotFoundException("Marker not found", "40430"));

        for (News news : newsRepo.findAll()) {
            Set<Long> markerIds = news.getMarkerIds();
            if (markerIds != null && markerIds.contains(id)) {
                Set<Long> updatedIds = new HashSet<>(markerIds);
                updatedIds.remove(id);
                news.setMarkerIds(updatedIds);
                newsRepo.update(news.getId(), news);
            }
        }

        repo.deleteById(marker.getId());
    }

    private void ensureNameUnique(String name, Long selfId) {
        boolean exists = repo.findAll().stream()
                .anyMatch(marker ->
                        marker.getName().equalsIgnoreCase(name)
                                && (selfId == null || !marker.getId().equals(selfId)));

        if (exists) {
            throw new ConflictException("Marker name must be unique", "40930");
        }
    }
}

File: 351003/Guzaev/publisher/app/services/marker_service.py
from repositories.base_repository import BaseRepository
from models.models import Marker
from dtos.marker_dto import MarkerRequestTo, MarkerResponseTo
from database import SessionLocal
from errors import AppError
from typing import List

marker_repo = BaseRepository[Marker](Marker)


class MarkerService:
    def create(self, dto: MarkerRequestTo) -> MarkerResponseTo:
        with SessionLocal() as db:
            existing = db.query(Marker).filter(Marker.name == dto.name).first()
            if existing:
                raise AppError(status_code=403, message="Marker name already exists", error_code=40307)

            entity = Marker(name=dto.name)
            saved = marker_repo.create(db, entity)
            return MarkerResponseTo(id=saved.id, name=saved.name)

    def get_all(self) -> List[MarkerResponseTo]:
        with SessionLocal() as db:
            entities = marker_repo.get_all(db)
            return [MarkerResponseTo(id=e.id, name=e.name) for e in entities]

    def get_by_id(self, id: int) -> MarkerResponseTo:
        with SessionLocal() as db:
            entity = marker_repo.get_by_id(db, id)
            if not entity:
                raise AppError(status_code=404, message="Marker not found", error_code=40407)
            return MarkerResponseTo(id=entity.id, name=entity.name)

    def update(self, id: int, dto: MarkerRequestTo) -> MarkerResponseTo:
        with SessionLocal() as db:
            entity = marker_repo.get_by_id(db, id)
            if not entity:
                raise AppError(status_code=404, message="Marker not found", error_code=40408)

            existing = db.query(Marker).filter(Marker.name == dto.name, Marker.id != id).first()
            if existing:
                raise AppError(status_code=403, message="Marker name already exists", error_code=40308)

            entity.name = dto.name
            db.commit()
            db.refresh(entity)
            return MarkerResponseTo(id=entity.id, name=entity.name)

    def delete(self, id: int) -> None:
        with SessionLocal() as db:
            success = marker_repo.delete(db, id)
            if not success:
                raise AppError(status_code=404, message="Marker not found", error_code=40409)
